ffi_internals/src/alias_resolution.rs

```diff
@@ -0,0 +1,311 @@
+//!
+//! This module provides methods for parsing aliases out of a module and writing their definitions
+//! to disk, so that when ffi_* macros encounter references to aliases defined in remote crates, the
+//! macros can identify the underlying type so they can determine safe FFI behavior. For example, if
+//! `CrateA` defines
+//! ```ignore
+//! type Foo = u16
+//! ```
+//! and `CrateB` references it
+//! ```ignore
+//! use CrateA::Foo;
+//!
+//! struct Bar { foo: Foo }
+//! ```
+//!
+//! In order to `ffi_derive::FFI`, we need to be able to determine that `Foo` is `u16` and should be
+//! exposed following the same rules as `u16`. This requires us to
+//! # Parse the alias definition into data we can work with in procedural macros.
+//! # Store that data, in a format that support storage of multiple alias sources.
+//! # Read that data while deriving the FFI for types in any other crate.
+//!
+
+use lazy_static::lazy_static;
+use quote::format_ident;
+use std::{collections::HashMap, io::BufRead, sync::Mutex};
+use syn::{Attribute, Ident, Item, ItemMod, ItemType, Lit, Meta::NameValue, Type};
+
+lazy_static! {
+    /// The path to the alias map file, behind a `Mutex` to ensure that multiple operations don't
+    /// attempt to write to it at once (which could result in a corrupted data structure).
+    ///
+    /// This is only an issue for tests since they're executed in parallel; rustc doesn't currently
+    /// do any parallel compilation. Still better to be safe and be able to test it, though.
+    ///
+    static ref ALIAS_MAP_PATH: Mutex<String> = Mutex::new(format!("{}/alias_map.json", env!("OUT_DIR")));
+}
+
+/// Describes the data for a type alias.
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
+pub(super) struct AliasDefinition {
+    /// The type that a newtype is defined as. In `type Foo = u16`, this is `u16`.
+    pub definition: String,
+    /// Set if `definition` is itself an alias, so that we can look at the outer keys again.
+    pub definition_source: Option<String>,
+}
+
+/// Parses `module` to create a hashmap of alias definitions so that we can resolve aliases to their
+/// underlying types when deriving the FFI.
+///
+pub fn parse_alias_module(resolution_key: String, module: ItemMod) -> ItemMod {
+    // Parse the alias resolution data out of `module`.
+    let (brace, items) = module
+        .clone()
+        .content
+        .unwrap_or_else(|| panic!("No module content? {:?}", module));
+    let (stripped_items, new_aliases): (Vec<Item>, HashMap<String, AliasDefinition>) = items
```

Comment:
Maybe NBD since I think you're only using this locally, but recent experience with libcompliance has me steering away from tuples with complex types. Maybe your code would get easier to reason about if you had:

```rust
struct Things {
  stripped_items: Vec<Item>,
  new_aliases: HashMap<String, AliasDefintion>
}
```

If you did a `#[derive(Default)]`, you could then also do a `Things::default()` that you pass into `fold()` below. Again, not a huge change, but maybe just makes it a tad easier to read.

Reply:
Thanks, that's a lot easier to look at